src/ZODB/tests/racetest.py

```diff
@@ -0,0 +1,515 @@
+from ZODB.tests.MinPO import MinPO
+from ZODB.tests.util import with_high_concurrency
+
+import threading
+from random import randint
+
+
+class ISpec:
+    """ISpec interface represents testing specification used by check_race_*"""
+
+    def init(root):
+        """init should initialize database state."""
+
+    def next(root):
+        """next should modify database state."""
+
+    def assertStateOK(root):
+        """assertStateOK should verify whether database state follows
+        intended invariant.
+
+        If not - it should raise AssertionError with details.
+        """
+
+
+class T2ObjectsInc(ISpec):
+    """T2ObjectsInc is specification with behaviour where two objects obj1
+    and obj2 are incremented synchronously.
+
+    It is used in tests where bugs can be immedeately observed after the race.
+
+    invariant:  obj1 == obj2
+    """
+    def init(_, root):
+        root['obj1'] = MinPO(0)
+        root['obj2'] = MinPO(0)
+
+    def next(_, root):
+        root['obj1'].value += 1
+        root['obj2'].value += 1
+
+    def assertStateOK(_, root):
+        # both objects must have the same values
+        i1 = root['obj1'].value
+        i2 = root['obj2'].value
+
+        if not (i1 == i2):
+            raise AssertionError("obj1 (%d)  !=  obj2 (%d)" % (i1, i2))
+
+
+class T2ObjectsInc2Phase(ISpec):
+    """T2ObjectsInc2Phase is specification with behaviour where two objects
+    obj1 and obj2 are incremented in lock-step.
+
+    It is used in tests where bugs can be observed on the next transaction
+    after the race.
+
+    invariant:  obj1 - obj2 == phase
+    """
+    def init(_, root):
+        root['obj1'] = MinPO(0)
+        root['obj2'] = MinPO(0)
+        root['phase'] = MinPO(0)
+
+    def next(_, root):
+        phase = root['phase']
+        if phase.value == 0:
+            root['obj1'].value += 1
+        else:
+            root['obj2'].value += 1
+        phase.value += 1
+        phase.value %= 2
+
+    def assertStateOK(_, root):
+        i1 = root['obj1'].value
+        i2 = root['obj2'].value
+        p = root['phase'].value
+
+        if not (i1 - i2 == p):
+            raise AssertionError("obj1 (%d) - obj2(%d) != phase (%d)" %
+                                 (i1, i2, p))
+
+
+class RaceTests(object):
+
+    # verify storage/Connection for race in between load/open and local
+    # invalidations.
+    # https://github.com/zopefoundation/ZEO/issues/166
+    # https://github.com/zopefoundation/ZODB/issues/290
+    def check_race_loadopen_vs_local_invalidate(self):
+        return self._check_race_loadopen_vs_local_invalidate(T2ObjectsInc())
+
+    @with_high_concurrency
+    def _check_race_loadopen_vs_local_invalidate(self, spec):
+        assert isinstance(spec, ISpec)
+        db = DB(self._storage)
+
+        # init initializes the database according to the spec.
+        def init():
+            _state_init(db, spec)
+
+        # verify accesses the database and verifies spec invariant.
```

Comment:
Thanks for the note here. Previously this text was

```py
        # verify accesses obj1/obj2 and verifies that obj1.value == obj2.value
        #
        # access to obj1 is organized to always trigger loading from zstor.
        # access to obj2 goes through zconn cache and so verifies whether the
        # cache is not stale.
```

but, as you note, it was not generlized well.

Do you think that it will be better e.g. as

```py
    # verify accesses objects in the database and verifies spec invariant.
    ...
```

?